### PR TITLE
Adding Tenant Config UI Feature Support

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
## Methodology

Fix for https://github.com/wso2/product-apim/issues/11973

## UI Implementation

Screenshot 2021-11-09 at 15 04 52

### CODE REVIEW

File: portals/admin/site/public/locales/en.json [hunk 1]
```diff
@@ -330,6 +331,10 @@
   "RolePermissions.TreeView.PermissionsSelector.scope.assignment.title": "Select Scope Assignments",
   "RolePermissions.TreeView.PermissionsSelector.update.scope.error": "Something went wrong while updating the permission",
   "ScopeAssignments.List.search.default": "Search by Role Name",
+  "Settings.Advanced.TenantConf.edit.success": "Tenant Configuration edited successfully",
```
Comment (bhathiya) [inline portals/admin/site/public/locales/en.json:334]: shouldn't it be "saved" instead?

Comment (CrowleyRajapakse) [reply portals/admin/site/public/locales/en.json:334]: ack, will change to saved

Comment (CrowleyRajapakse) [reply portals/admin/site/public/locales/en.json]: resolved

File: portals/admin/site/public/locales/en.json [hunk 2]
```diff
@@ -330,6 +331,10 @@
   "RolePermissions.TreeView.PermissionsSelector.scope.assignment.title": "Select Scope Assignments",
   "RolePermissions.TreeView.PermissionsSelector.update.scope.error": "Something went wrong while updating the permission",
   "ScopeAssignments.List.search.default": "Search by Role Name",
+  "Settings.Advanced.TenantConf.edit.success": "Tenant Configuration saved successfully",
```
Comment (bhathiya) [inline portals/admin/site/public/locales/en.json:334]: Is this the message we show once saved? Wondering why we have "Tenant" only here. In other places we have "Advanced" right?

Comment (CrowleyRajapakse) [reply portals/admin/site/public/locales/en.json:334]: ahh, yes, will change to Advanced to be consistent with other places

Comment (CrowleyRajapakse) [reply portals/admin/site/public/locales/en.json]: resolved

**Comment by CrowleyRajapakse** [comment]
Latest UI with Schema validation + menu name change

Screenshot 2021-11-11 at 16 08 55

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+/**
+ * Render a list
+ * @returns {JSX} Header AppBar components.
+ */
+function TenantConfSave(props) {
+    const classes = useStyles();
+    const {
+        updateList, history,
```
Comment (tmkasun) [inline portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:97]: shall we use `useHistory` from react-router, it seems like parent component is passing it's history object to this component

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    } = props;
+    const intl = useIntl();
+    const [initialState, setInitialState] = useState({
```
Comment (tmkasun) [inline portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:100]: We don't need a separate initial state for the reducer and this is not working as intended, Reseting the state should be done via the dispatcher it self,
example
```js

function reducer(state, action) {
  switch (action.type) {
    case 'increment':
      return {count: state.count + 1};
    case 'decrement':
      return {count: state.count - 1};
    case 'reset':
      return {tenantConf: '', tenantConfSchema: ''};
    default:
      throw new Error();
  }
}

function Counter({initialCount}) {
  const [state, dispatch] = useReducer(reducer, initialValue);
  return (
    <>
      Count: {state.count}
      <button
        onClick={() => dispatch({type: 'reset', payload: initialCount})}>
        Reset
      </button>
      <button onClick={() => dispatch({type: 'decrement'})}>-</button>
      <button onClick={() => dispatch({type: 'increment'})}>+</button>
    </>
  );
}

```

docs: https://reactjs.org/docs/hooks-reference.html

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        let tenantConfSchemaVal;
+        restApi.tenantConfSchemaGet().then((result) => {
+            tenantConfSchemaVal = result.body;
+            dispatch({ field: 'tenantConfSchema', value: tenantConfSchemaVal });
+        });
+        restApi.tenantConfGet().then((result) => {
+            tenantConfVal = JSON.stringify(result.body, null, '\t');
+            dispatch({ field: 'tenantConf', value: tenantConfVal });
+        });
+        setInitialState({
```
Comment (tmkasun) [inline portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:121]: This has no effect on the reducer's initial state

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+    const cancelUpdate = () => {
+        let tenantConfVal;
+        restApi.tenantConfGet().then((result) => {
+            tenantConfVal = JSON.stringify(result.body, null, '\t');
+            const editState = {
+                tenantConf: tenantConfVal,
+            };
+            dispatch({ field: 'editDetails', value: editState });
+        });
+        setInitialState({
```
Comment (tmkasun) [inline portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:154]: Same as above have no effect on the reducer's initial state, Have to use the dispatcher instead 

is this `cancelUpdate` work as expected? is it reset the reducer's state to initial value ?

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                    />,
+                );
+                history.push('/settings/advanced');
+            })
+            .catch((error) => {
+                const { response } = error;
+                if (response.body) {
+                    Alert.error(response.body.description);
+                }
+                return null;
```
Comment (tmkasun) [inline portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:177]: Log the error please

```js
console.error(error)
```
and why do we need to return in the catch block?

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+            })
+            .finally(() => {
+                updateList();
```
Comment (tmkasun) [inline portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:180]: Don't do this kind of work in finally, If something goes wrong in updateList function we are clue less, do only finalization task such as stopping an animation (loaders) , cancel a network request ect

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+            });
+    };
+
+    return (
+        <ContentBase
+            pageStyle='full'
+            title={intl.formatMessage({
```
Comment (tmkasun) [inline portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:187]: Should have use `<FormattedMessage/>` component

**Review by tmkasun — CHANGES_REQUESTED** [review]

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 3 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx]: fixed

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 7 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:180]: resolved

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 8 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx]: resolved

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 9 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx]: resolved

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 6 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:155]: yes, currently it reset to initial value without issues

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 4 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:101]: will change implementation to proper way

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 5 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:122]: will change accordingly

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 6 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx:155]: will change accordingly

Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx]: resolved

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 5 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx]: resolved

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 4 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx]: for this line const [state, dispatch] = useReducer(reducer, initialValue); it seems we need to initial value, as per the doc also having initial value is okay and only if we need lazy initialisation we don't need to have initial value it seems and need to provide a method instead, so I removed only the unnecessary use of setInitialState method and related usages.

File: portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx (hunk 7 above)
Comment (CrowleyRajapakse) [reply portals/admin/source/src/app/components/AdvancedSettings/TenantConfSave.jsx]: removed the return statement as well

**Review by tmkasun — APPROVED** [review]

**Pull request merged**